Allow `experimental_shell_command`/`experimental_run_in_sandbox` to specify `output_`s from anywhere under the buildroot

chrisjrn opened the pull request:
Over in #17928, @danxmoran slightly nerdsniped me :)

WIth this change, ad-hoc processes are wrapped by a `bash` process that runs in the buildroot, and immediately `chdir`s into the working directory, rather than using `Process`' native `working_directory` argument. This means that Pants will allow the processes to request output files from anywhere under the buildroot, rather than just under the working directory.

Prework for #17928, addresses one of the use cases in #16807.

kaos commented on src/python/pants/backend/shell/util_rules/shell_command.py:
```diff
@@ -488,6 +486,37 @@ async def prepare_shell_command_process(
+
+    working_directory = process.working_directory
+    if working_directory is None:
+        working_directory = ""
```
saving some line-noise, this could just be:
```python
working_directory = process.working_directory or ""
```

kaos commented on src/python/pants/backend/shell/util_rules/shell_command_test.py:
```diff
@@ -553,7 +555,8 @@ def test_shell_command_boot_script(rule_runner: RuleRunner) -> None:
         "-c",
-        (
+        "cd src && /bin/bash -c "
+        + shlex.quote(
```
I'd like to also see a test without the `cd` prefix.. (as there'll be a superfluous space in there I'd like to get rid of :) )

chrisjrn replied:
that's a good idea anyway, imho :)

kaos reviewed (approved):
Neat 👍🏽

chrisjrn commented:
@kaos I've addressed your two concerns in https://github.com/pantsbuild/pants/pull/17938/commits/c068e39d7042fc267391f3611a6a3e3c0e0097d1 and also added some better documentation in https://github.com/pantsbuild/pants/pull/17938/commits/2cf8830c7707de176e5848cd17966a824605dde4

kaos reviewed (approved):
Nice 👍🏽
